src/borg/remote.py

```diff
@@ -715,6 +717,10 @@ def handle_error(unpacked):
         calls = list(calls)
         waiting_for = []
         while wait or calls:
+            if self.shutdown_time and time.time() > self.shutdown_time:
+                # we are shutting this RemoteRepository down already, make sure we do not waste
+                # a lot of time in case a lot of async stuff is coming in or remote is gone or slow.
+                return
```

Comment:
Add a debug log that includes len(waiting_for) and len(self.async_responses) here?